### DISCUSSION TITLE
Capacitor 3: AndroidX Activity Result API

### DISCUSSION THREAD

**Discussion by AE1NS** [ionic-team/capacitor#4704] [Q&A] [answered]
Hello,

I have problems with refactoring my google pay plugin to capacitor 3. Depending to [Android Request Codes](https://capacitorjs.com/docs/updating/plugins/3-0#android-request-codes) I tried to refactor my code and removed the 'requestCodes' attribute and changed my 'handleOnActivityResult' to a private method with the @ActivityCallback annotation. So far so good.

To start the google pay intent I use this helper: [AutoResolveHelper](https://developers.google.com/android/reference/com/google/android/gms/wallet/AutoResolveHelper) that gets a resultCode int as parameter. So am I right that this needs to be refactored to a 'startActivityForResult'? Or is the AutoResolveHelper not updated to this 'new' callbackname parameter? Does someone know a solution for this?

**Comment by DavidStrausz** [comment, marked as answer]
I may be a bit late with my answer, but I ran into the same problem today - maybe it's useful for somebody.

I was able to work around all deprecations by not using `AutoResolveHelper` or `startActivityFoResult` at all. Instead, I register an `onFailureListener` for the payment request task manually, and handle the exception (which conveniently is a [ResolvableApiException](https://developers.google.com/android/reference/com/google/android/gms/common/api/ResolvableApiException)) myself:

```java
Task<PaymentData> paymentResultTask = paymentsClient.loadPaymentData(request);
paymentResultTask.addOnFailureListener(getActivity(), exception -> {
    PendingIntent pendingIntent = ((ResolvableApiException) exception).getResolution();
    IntentSenderRequest intentSenderRequest = new IntentSenderRequest.Builder(pendingIntent.getIntentSender())
            .build();

    gPayActivityResultHandler.launch(intentSenderRequest);
});
```

For this to work you need to register an AndroidX [ActivityResultHandler](https://developer.android.com/training/basics/intents/result) (you need to do this in your plugins' `load` method):

```java
gPayActivityResultHandler = bridge.registerForActivityResult(
        new ActivityResultContracts.StartIntentSenderForResult(),
        (ActivityResultCallback<ActivityResult>) result -> {
            switch (result.getResultCode()) {
                case Activity.RESULT_OK:
                    Intent data = result.getData();
                    PaymentData paymentData = PaymentData.getFromIntent(data);
                    ...
                    break;
                case Activity.RESULT_CANCELED:
                    ...
                    break;
            }
        });
```

**Comment by AE1NS** [comment]
I just successfully implemented this, thank you!

In my case I have one method in my capacitor plugin called 'loadPaymentData'. I use this.saveCall(call); before I call the native 'loadPaymentData'. When the handler 'bridge.registerForActivityResult' is called, I read the call again with 'PluginCall call = this.getSavedCall();'. With this implementation, my plugins method resolves, when the google pay dialog closes.
Unfortunately 'saveCall' and 'getSavedCall' is also deprecated. The only idea I have right now, is to create a plugin event, but it doesnt seem comfortable to me to call the plugins 'loadPaymentData' and get the result in a plugins event listener.
Do you also have any solution for this issue?
